models/trie_mimc/frontend_proof.go

```diff
@@ -0,0 +1,477 @@
+package trie_mimc
+
+import (
+	"github.com/consensys/gnark/frontend"
+	"github.com/consensys/gnark/std/hash/mimc"
+)
+
+// Return the result of right shift by 1, input size = 32
+// Input: {42, 194, 10, 96, 133, 113, 88, 31, 86, 136, 60, 65, 11, 106, 226, 218, 169, 220, 186, 36, 114, 230, 53, 147, 171, 202, 12, 106, 45, 89, 231, 132}
+// Output: {42, 194, 10, 96, 133, 113, 88, 31, 86, 136, 60, 65, 11, 106, 226, 218, 169, 220, 186, 36, 114, 230, 53, 147, 171, 202, 12, 106, 45, 89, 231}
+func rightShift1(api frontend.API, input frontend.Variable) frontend.Variable {
+	var lsb8 frontend.Variable = 0
+	var multiplier frontend.Variable = 1
+	inputBinary := api.ToBinary(input)
+	for i := 0; i < 8; i++ {
+		lsb8 = api.Add(lsb8, api.Mul(inputBinary[i], multiplier))
+		multiplier = api.Mul(multiplier, 2)
+	}
+	return api.DivUnchecked(api.Sub(input, lsb8), 256)
```

Comment:
the reason we have to substract 8lsb is that DivUnchecked does not remove the remainder? 
what would be the result if you just used `api.DivUnchecked(input, 256)` ?

Reply:
This is a good point.  DivUnchecked does not remove the remainder. I also thought it can remove the remainder but it cannot. Only if those remainders are 0 then we can divided it. In, short [A, B] / 256 != [0, A], unless B == 0.

Reply:
To find out the logic I tried the followings before. Maybe it helps:
```
A = []byte{41, 110, 104, 212, 140, 165, 49, 87, 44, 32, 181, 221, 58, 86, 232, 118, 168, 2, 169, 90, 13, 92, 128, 54, 250, 117, 253, 216, 155, 33, 211, 90} (32 bytes)
api.Println(A) = 18739903480979981037871518428324519730095599429594896857256085181235466457946
api.Println(api.DivUnchecked(A, 256)) = 14266360235180858077854589344300906970426265976166627335590460859971951862228
B = []byte{0, 41, 110, 104, 212, 140, 165, 49, 87, 44, 32, 181, 221, 58, 86, 232, 118, 168, 2, 169, 90, 13, 92, 128, 54, 250, 117, 253, 216, 155, 33, 211} (32 bytes)
api.Println(B) = 73202747972578050929185618860642655195685935271855065848656582739201040851
```
As you can see api.DivUnchecked(A, 256) != B

Reply:
did you check what the []byte{..} of (api.DivUnchecked(A,256)) looks like?

Reply:
It will need to dig into the function implementation (like the followings)
```
func (system *r1cs) DivUnchecked(i1, i2 frontend.Variable) frontend.Variable {
	vars, _ := system.toVariables(i1, i2)

	v1 := vars[0]
	v2 := vars[1]

	n1, v1Constant := system.ConstantValue(v1)
	n2, v2Constant := system.ConstantValue(v2)

	if !v2Constant {
		res := system.newInternalVariable()
		debug := system.AddDebugInfo("div", v1, "/", v2, " == ", res)
		// note that here we don't ensure that divisor is != 0
		system.addConstraint(newR1C(v2, res, v1), debug)
		return res
	}

	// v2 is constant
	if n2.IsUint64() && n2.Uint64() == 0 {
		panic("div by constant(0)")
	}
	q := system.CurveID.Info().Fr.Modulus()
	n2.ModInverse(n2, q)

	if v1Constant {
		n2.Mul(n2, n1).Mod(n2, q)
		return system.toVariable(n2)
	}

	// v1 is not constant
	return system.mulConstant(v1, system.toVariable(n2).(compiled.LinearExpression))
}

// Div returns res = i1 / i2
func (system *r1cs) Div(i1, i2 frontend.Variable) frontend.Variable {
	vars, _ := system.toVariables(i1, i2)

	v1 := vars[0]
	v2 := vars[1]

	n1, v1Constant := system.ConstantValue(v1)
	n2, v2Constant := system.ConstantValue(v2)

	if !v2Constant {
		res := system.newInternalVariable()
		debug := system.AddDebugInfo("div", v1, "/", v2, " == ", res)
		v2Inv := system.newInternalVariable()
		// note that here we ensure that v2 can't be 0, but it costs us one extra constraint
		system.addConstraint(newR1C(v2, v2Inv, system.one()), debug)
		system.addConstraint(newR1C(v1, v2Inv, res), debug)
		return res
	}

	// v2 is constant
	if n2.IsUint64() && n2.Uint64() == 0 {
		panic("div by constant(0)")
	}
	q := system.CurveID.Info().Fr.Modulus()
	n2.ModInverse(n2, q)

	if v1Constant {
		n2.Mul(n2, n1).Mod(n2, q)
		return system.toVariable(n2)
	}

	// v1 is not constant
	return system.mulConstant(v1, system.toVariable(n2).(compiled.LinearExpression))
}
```
Unfortunately I haven't had time to dig into it.

Reply:
The bytes cannot be shown directly in gnark. Will need to implement a tool to translate the big number to show that.

Reply:
ok thx. could be that the div is defined on the field instead of integers